src/interfaces/txmempool.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <interfaces/txmempool.h>
+
+#include <primitives/transaction.h>
+#include <sync.h>
+#include <txmempool.h>
+#include <util/memory.h>
+
+#include <memory>
+#include <vector>
+
+namespace interfaces {
+namespace {
+
+class TxMempoolImpl final : public TxMempool
+{
+public:
+    explicit TxMempoolImpl(CTxMemPool& txmempool)
+        : TxMempool{&txmempool.cs},
+          m_txmempool(txmempool) {}
+
+    void transactionUpdated() override
+    {
+        m_txmempool.AddTransactionsUpdated(1);
+    }
+
+    void check(const CCoinsViewCache& coins) override
+    {
+        m_txmempool.check(&coins);
+    }
+
+    void removeForBlock(const std::vector<CTransactionRef>& vtx, unsigned int block_height) override EXCLUSIVE_LOCKS_REQUIRED(m_mutex)
```

Comment:
```suggestion
    void removeForBlock(const std::vector<CTransactionRef>& vtx, unsigned int block_height) override EXCLUSIVE_LOCKS_REQUIRED(m_txmempool.cs)
```

Reply:
Annotation can not be added to cpp files (or at least they will be ignored). Moved annotation to header file.

Reply:
> Annotation can not be added to cpp files (or at least they will be ignored).

I don't think so.
https://clang.llvm.org/docs/ThreadSafetyAnalysis.html:
> Attributes must be attached to named declarations, such as classes, methods, and data members.

Reply:
Just logically, if validation does not include the txmempool.cpp file, how would it know about the annotation?

And practically you observed that it didn't work: https://github.com/bitcoin/bitcoin/pull/19629#issuecomment-674423662